File: 0001_two_sum/two_sum.cpp
#include <vector>
#include <unordered_map>

using namespace std;

vector<int> two_sum(vector<int>& nums, int target) {
    unordered_map<int, int> needed;
    for (int i = 0; i < nums.size(); i++) {
        int n = nums[i];
        if (needed.count(target - n))
            return {needed[target - n], i};
        else
            needed[n] = i;
    }
    return {}; // Should never come to this.
}

int main () {}

File: two_sum/two_sum.py
# We want to find n and m such that n + m = target.
# Keep track of numbers we've seen so far and their index.
# For each number n, check to see if m is in the dict.
# Using a dict makes this O(n).
def twoSum(self, nums: List[int], target: int) -> List[int]:
    needed = {}
    for i, n in enumerate(nums):
        if (target - n) in needed:
            return [needed[target - n], i]
        else:
            needed[n] = i
